extensions/wikia/WikiFactory/SpecialWikiFactory_body.php

```diff
@@ -61,160 +65,201 @@ public function execute( $subpage ) {
-					$cityid = WikiFactory::DBtoID( strtolower( $name ) );
-				}
-				elseif( preg_match( "/^id\:/", $domain ) ) {
-					list( $prefix, $cityid ) = explode( ":", $domain, 2 );
-				}
-				else {
-					/**
-					 * if there's no dot in cityname we add .wikia.com
-					 * or if is only one dot (language.domain.wikia.com)
-					 */
-					if( sizeof(explode(".", $domain )) <= 2 && strlen( $domain ) > 0) {
-						$this->mDomain = $domain;
-						$domain = $domain.".".$wgWikiaBaseDomain;
-					} else {
-						list( $code, $subdomain ) = explode(".", $domain, 2 );
-						$exists = 0;
-						if ( in_array( $code, array_keys( Language::getLanguageNames() ) ) ) {
-							$exists = 1;
-						} else {
-							$cityid = WikiFactory::DomainToId( $domain );
-							if ( !is_null($cityid) ) {
-								$exists = 1;
-							}
-						}
-						if ( $exists == 1 ) {
-							$this->mDomain = $code.".".$subdomain;
-							$domain = sprintf("%s.%s", $code, $subdomain );
-						}
-					}
-					if ( is_null($cityid) ) {
-						$cityid = WikiFactory::DomainToId( $domain );
-					}
-				}
+			if ( ( $this->mTab === "tags" || $this->mTab === "findtags" )
+			     && ( isset( $parts[2] ) ) ) {
+				$this->mSearchTag = trim( $parts[2] );
 			}
 		}
-		$this->mTab = $tab;
-		if( !is_null( $cityid ) ) {
-			$this->mTitle = Title::makeTitle( NS_SPECIAL, "WikiFactory/{$cityid}/{$tab}" );
+	}
+
+	/**
+	 * The City ID can be passed in one of several locations:
+	 *
+	 *   - The "cityId" query parameter
+	 *   - The first subpage component
+	 *   - The "wpCityDomain" query parameter if it has the format "id:1234"
+	 *
+	 * @param string $subpage The parts of the page after Special:WikiFactory
+	 *
+	 * @return String
+	 */
+	private function findCityId( $subpage ) {
+		// Look in the cityId query parameter first
+		$cityId	= F::app()->wg->Request->getVal( "cityId", null );
+		if ( !empty( $cityId ) ) {
+			return $cityId;
 		}
-		if ( !isset($this->mVariableName) ) {
-			$this->mVariableName = "";
+
+		$pathParts = explode( '/', $subpage );
+		if ( !empty( $pathParts[0] ) && ctype_digit( $pathParts[0] ) ) {
+			return $pathParts[0];
 		}
-		return WikiFactory::getWikiByID( $cityid );
+
+		return $this->findCityIdFromDomain( $subpage );
 	}
 
 	/**
-	 * doWikiSelector
+	 * Determine the cityId by looking for a domain.  The domain is found in the
+	 * wpCityDomain parameter and can have any of the following formats
 	 *
-	 * Pager with all wikis
+	 * - id:1234 : A city ID masquerading as a domain
+	 * - db:Callofduty : A dbName
+	 * - callofduty : For single names (no dot separators) add .wikia.com and try as domain
+	 * - callofduty.wikia.com : A full domain
+	 * - either of the previous two used in the subdomain
+	 * - http://callofduty.wikia.com/wiki/Page : Any pasted URL
+	 *
+	 * @param string $subpage
 	 *
-	 * @access private
-	 * @author eloy
+	 * @return mixed
+	 *
+	 * @throws Exception
+	 */
+	private function findCityIdFromDomain( $subpage ) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce the number of returns of this function 5, down to the maximum allowed 3. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1142)